From d3d693fddcb73a9b360b8bdf47b3661eed84cd1c Mon Sep 17 00:00:00 2001
From: Thomas Haller <thaller@redhat.com>
Date: Tue, 12 Mar 2024 10:10:00 +0100
Subject: [PATCH 6/8] pangofc-font: rework wrong usage of GWeakRef in
 PangoFcFont

PangoFcFont has a field `PangoFcFontMap *fontmap`, but at various
places, this field is cast to a `GWeakRef` and used with g_weak_ref_*()
API.

Correctly, all places that set the field, would do so via
g_weak_ref_set(). And "accidentally", sizeof(GWeakRef) is the same as
`sizeof(gpointer)`. The wrong usage happens at the places that access
the fontmap field directly, and assume this is just a plain pointer.

This mostly works. Note that [1] reworked the representation, so while
another thread accesses the GWeakRef (e.g. by destroying the fontmap),
the pointer gets mangled. If the code would then just access the fontmap
field, it would be mangled. Probably that does not happen, and pango
objects are not unreffed or released on other threads. So it somewhat
worked.

But it also relies on internals of glib and (since [1]) it relies that
the object is not destroyed on another thread.

Rework that. Instead of wrongly using a GWeakRef, use
g_object_add_weak_pointer(). That is also not thread-safe, but that
doesn't seem to be a requirement here anyway.

Note that PangoWin32Font has a similar problem. But it's more severe
because it directly assigns pointers instead of using g_weak_ref_set()
(see _pango_cairo_win32_font_new()). That leads to crashes since [1].
This will be fixed next.

[1] https://gitlab.gnome.org/GNOME/glib/-/merge_requests/3834

Fixes: aca468cdc2ea ('Port pangofc-font.c to GWeakRef')
(cherry picked from commit 8e8f894da8f3ef5486aa07a02db56827966232dc)
---
 pango/pangofc-font.c | 40 +++++++++++++++-------------------------
 pango/pangofc-font.h |  3 ++-
 2 files changed, 17 insertions(+), 26 deletions(-)

diff --git a/pango/pangofc-font.c b/pango/pangofc-font.c
index 053e93cd..6d38eb49 100644
--- a/pango/pangofc-font.c
+++ b/pango/pangofc-font.c
@@ -162,17 +162,16 @@ pango_fc_font_finalize (GObject *object)
 {
   PangoFcFont *fcfont = PANGO_FC_FONT (object);
   PangoFcFontPrivate *priv = fcfont->priv;
-  PangoFcFontMap *fontmap;
+  PangoFontMap *fontmap;
 
   g_slist_foreach (fcfont->metrics_by_lang, (GFunc)free_metrics_info, NULL);
   g_slist_free (fcfont->metrics_by_lang);
 
-  fontmap = g_weak_ref_get ((GWeakRef *) &fcfont->fontmap);
+  fontmap = fcfont->fontmap;
   if (fontmap)
     {
-      _pango_fc_font_map_remove (PANGO_FC_FONT_MAP (fcfont->fontmap), fcfont);
-      g_weak_ref_clear ((GWeakRef *) &fcfont->fontmap);
-      g_object_unref (fontmap);
+      g_object_remove_weak_pointer (G_OBJECT (fontmap), (gpointer *) &fcfont->fontmap);
+      _pango_fc_font_map_remove (PANGO_FC_FONT_MAP (fontmap), fcfont);
     }
 
   pango_font_description_free (fcfont->description);
@@ -240,7 +239,10 @@ pango_fc_font_set_property (GObject       *object,
         PangoFcFontMap *fcfontmap = PANGO_FC_FONT_MAP (g_value_get_object (value));
 
         g_return_if_fail (fcfont->fontmap == NULL);
-        g_weak_ref_set ((GWeakRef *) &fcfont->fontmap, fcfontmap);
+
+        fcfont->fontmap = (PangoFontMap *) fcfontmap;
+        if (fcfontmap)
+          g_object_add_weak_pointer (G_OBJECT (fcfontmap), (gpointer *) &fcfont->fontmap);
       }
       goto set_decoder;
 
@@ -274,8 +276,8 @@ pango_fc_font_get_property (GObject       *object,
     case PROP_FONTMAP:
       {
 	PangoFcFont *fcfont = PANGO_FC_FONT (object);
-	PangoFontMap *fontmap = g_weak_ref_get ((GWeakRef *) &fcfont->fontmap);
-	g_value_take_object (value, fontmap);
+
+	g_value_set_object (value, fcfont->fontmap);
       }
       break;
     default:
@@ -331,8 +333,6 @@ pango_fc_font_get_coverage (PangoFont     *font,
   PangoFcFont *fcfont = (PangoFcFont *)font;
   PangoFcFontPrivate *priv = fcfont->priv;
   FcCharSet *charset;
-  PangoFcFontMap *fontmap;
-  PangoCoverage *coverage;
 
   if (priv->decoder)
     {
@@ -340,13 +340,10 @@ pango_fc_font_get_coverage (PangoFont     *font,
       return _pango_fc_font_map_fc_to_coverage (charset);
     }
 
-  fontmap = g_weak_ref_get ((GWeakRef *) &fcfont->fontmap);
-  if (!fontmap)
+  if (!fcfont->fontmap)
     return pango_coverage_new ();
 
-  coverage = _pango_fc_font_map_get_coverage (fontmap, fcfont);
-  g_object_unref (fontmap);
-  return coverage;
+  return _pango_fc_font_map_get_coverage (PANGO_FC_FONT_MAP (fcfont->fontmap), fcfont);
 }
 
 /* For Xft, it would be slightly more efficient to simply to
@@ -477,7 +474,7 @@ pango_fc_font_get_metrics (PangoFont     *font,
       PangoFontMap *fontmap;
       PangoContext *context;
 
-      fontmap = g_weak_ref_get ((GWeakRef *) &fcfont->fontmap);
+      fontmap = fcfont->fontmap;
       if (!fontmap)
 	return pango_font_metrics_new ();
 
@@ -526,7 +523,6 @@ pango_fc_font_get_metrics (PangoFont     *font,
         }
 
       g_object_unref (context);
-      g_object_unref (fontmap);
     }
 
   return pango_font_metrics_ref (info->metrics);
@@ -1107,17 +1103,11 @@ static PangoLanguage **
 _pango_fc_font_get_languages (PangoFont *font)
 {
   PangoFcFont * fcfont = PANGO_FC_FONT (font);
-  PangoFcFontMap *fontmap;
-  PangoLanguage **languages;
 
-  fontmap = g_weak_ref_get ((GWeakRef *) &fcfont->fontmap);
-  if (!fontmap)
+  if (!fcfont->fontmap)
     return NULL;
 
-  languages  = _pango_fc_font_map_get_languages (fontmap, fcfont);
-  g_object_unref (fontmap);
-
-  return languages;
+  return _pango_fc_font_map_get_languages (PANGO_FC_FONT_MAP (fcfont->fontmap), fcfont);
 }
 
 /**
diff --git a/pango/pangofc-font.h b/pango/pangofc-font.h
index 25361fa0..9cde77e5 100644
--- a/pango/pangofc-font.h
+++ b/pango/pangofc-font.h
@@ -73,7 +73,8 @@ struct _PangoFcFont
   PangoFont parent_instance;
 
   FcPattern *font_pattern;          /* fully resolved pattern */
-  PangoFontMap *fontmap;            /* associated map */
+  PangoFontMap *fontmap;            /* associated map (no strong reference is held,
+                                     * but a g_object_add_weak_pointer() guards it) */
   gpointer priv;                    /* used internally */
   PangoMatrix matrix;               /* unused */
   PangoFontDescription *description;
-- 
2.47.0.windows.2

